File: src/browser.js
const {
    Builder,
    By,
    Key,
    until
} = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const proxy = require('selenium-webdriver/proxy');
const fs = require('fs');
const TokenValidator = require('./tokenValidator')
const player = require('./player')
const HttpsProxyAgent = require('https-proxy-agent');
const USER_AGENT = '--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36';

class Browser {
    constructor(userdata) {
        this.driver = null;
        this.userdata = userdata
    }



    async start() {

        const options = new chrome.Options();
        const proxydata = this.userdata.proxyAccount;
        if (proxydata) {

            

            const proxyUrl = `http://${proxydata.credentials}@${proxydata.host}`
            // const proxyUrl = 'http://H0vFDG:XoDcR2@45.92.22.6:8000';
            // options.addArguments(`--proxy-server=${proxyUrl}`); // proxyUrl - URL прокси сервера
            options.setProxy(proxy.manual({
                http:proxyUrl,
                https:proxyUrl
            }))
            // options.setProxy({
            //     proxyType: 'manual',
            //     httpProxy: proxyUrl,
            //     sslProxy: proxyUrl,
            //     noProxy: 'localhost'
            // }); // инициализируем https-proxy-agent
            // Создание Capabilities с использованием прокси
        }
        options.addArguments(USER_AGENT);
        options.addArguments('--disable-gpu');
        options.addArguments('--disable-blink-features=AutomationControlled');
        // options.addArguments('--headless');
        options.addArguments('--no-sandbox');
        options.addArguments('--disable-dev-shm-usage');
        options.addArguments('--disable-setuid-sandbox');
        options.addArguments('--window-size=1920,1080');
        this.driver = await new Builder()
            .forBrowser('chrome')
            .setChromeOptions(options)
            .build();


        await this.driver.get('https://spotify.com');




        const cookiesString = JSON.stringify(this.userdata.cookies);

        if (this.userdata.cookies) {
            const cookies = JSON.parse(cookiesString);
            for (const cookie of cookies) {
                await this.driver.manage().addCookie(cookie);
            }
        }

        
        this.driver.get('https://spotify.com');


        const session = await this.driver.wait(until.elementLocated(By.id('session')));

        if (! session) {
            return 401
        } else {
            await new Promise(resolve => setTimeout(resolve, Math.floor(Math.random() * 10000)));
            try {
                const closeButtonContainer = await this.driver.findElement(By.id('onetrust-close-btn-container'));
                if (closeButtonContainer) {
                  const closeButton = await closeButtonContainer.findElement(By.tagName('button'));
                  await closeButton.click();
                  await new Promise(resolve => setTimeout(resolve, Math.floor(Math.random() * 10000)));
                }
              } catch (error) {
                console.log(error);
                // Обработка ошибки
              }

            const playButton = await this.driver.wait(until.elementLocated(By.css('button[data-testid="play-button"]')));
            if (!playButton){
                return "error";
            }
            await this.driver.executeScript("arguments[0].click();", playButton);
            await new Promise(resolve => setTimeout(resolve, Math.floor(Math.random() * 10000)));

            let screenshot = await this.driver.takeScreenshot();
            fs.writeFileSync('screenshot.png', screenshot, 'base64');

        }
        return 200;
    }

    async navigate(url) {
        await this.driver.get(url);
    }

    async stop() {
        await this.driver.quit();
        this.driver = null;
    }
}

module.exports = Browser;

File: python/browser.py
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import os
import json
import base64

class Browser:
    def __init__(self, userdata):
        self.driver = None
        self.userdata = userdata

    def start(self):
        options = webdriver.ChromeOptions()
        proxydata = self.userdata['proxyAccount']
        # if proxydata:
        #     proxyUrl = 'http://{}@{}'.format(proxydata['credentials'], proxydata['host'])
        #     options.add_argument('--proxy-server={}'.format(proxyUrl))

        options.add_argument('--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 13_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36')
        options.add_argument('--disable-gpu')
        options.add_argument('--disable-blink-features=AutomationControlled')
        options.add_argument('--headless')
        options.add_argument('--no-sandbox')
        options.add_argument('--disable-dev-shm-usage')
        options.add_argument('--disable-setuid-sandbox')
        options.add_argument('--window-size=1920,1080')
        self.driver = webdriver.Chrome(ChromeDriverManager().install(), options=options)

        self.driver.get('https://spotify.com')

        cookies = self.userdata.get('cookies')
        if cookies:
            for cookie in cookies:
                self.driver.add_cookie(cookie)

        self.driver.get('https://spotify.com')

        try:
            session = WebDriverWait(self.driver, 10).until(EC.presence_of_element_located((By.ID, 'session')))
        except:
            return 401

        try:
            time.sleep(3)
            closeButtonContainer = self.driver.find_element(By.ID, 'onetrust-close-btn-container')
            if closeButtonContainer:
                closeButton = closeButtonContainer.find_element(By.TAG_NAME, 'button')
                closeButton.click()
                time.sleep(1)
        except:
            pass
        time.sleep(10)
        screenshot = self.driver.get_screenshot_as_base64()
        with open('screenshot.png', 'wb') as f:
            f.write(base64.b64decode(screenshot))
        
        playButton = WebDriverWait(self.driver, 10).until(EC.presence_of_element_located((By.CSS_SELECTOR, 'button[data-testid="play-button"]')))
        self.driver.execute_script("arguments[0].click();", playButton)
        time.sleep(10)

        # playButton = WebDriverWait(self.driver, 10).until(EC.presence_of_element_located((By.CSS_SELECTOR, 'button[data-testid="play-button"]')))
        # playButton.click()
        # time.sleep(1)

        screenshot = self.driver.get_screenshot_as_base64()
        with open('screenshot.png', 'wb') as f:
            f.write(base64.b64decode(screenshot))
        return 200

    def navigate(self, url):
        self.driver.get(url)

    def stop(self):
        self.driver.quit()
        self.driver = None

cookies = '[{"domain":".spotify.com","expiry":1715116100,"httpOnly":false,"name":"sp_t","path":"/","sameSite":"None","secure":true,"value":"f9e8ebedbf54625ca78df18a16a368b1"},{"domain":".spotify.com","expiry":1715116072,"httpOnly":false,"name":"sp_key","path":"/","sameSite":"None","secure":true,"value":"308f7bd5-bd80-46ff-8bce-9c3f2b274c93"},{"domain":".spotify.com","expiry":1683666473,"httpOnly":true,"name":"sp_landing","path":"/","sameSite":"None","secure":true,"value":"https%3A%2F%2Fopen.spotify.com%2F%3Fsp_cid%3Df9e8ebedbf54625ca78df18a16a368b1%26device%3Ddesktop"},{"domain":".spotify.com","expiry":1715116072,"httpOnly":true,"name":"sp_dc","path":"/","sameSite":"None","secure":true,"value":"AQCEl0hdWbwA84RIrzVgTr1McIrlk4ivTAs8AIIXabqddVd8stGk2ak5Edp04xzDWv6SCU63GRK0RQOSTcPnces4yVkFls2B4cuJNS60SqMg0tPoq5CS0ewrR-N8aN5w-7XPnhfP8KlCV9fILwmsNu-ia8_Ag3c"}]'

userdata = {
    'proxyAccount': {
        'credentials': '',
        'host': 'proxy.example.com'
    },
    'cookies': json.loads(cookies)
}
browser = Browser(userdata)
browser.start()
time.sleep(10)
# browser.navigate('https://example.com')
browser.stop()